File: src/main/java/site/gachontable/gachontablebe/global/config/aurora/CustomRoutingDataSource.java
package site.gachontable.gachontablebe.global.config.aurora;

import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource;
import org.springframework.transaction.support.TransactionSynchronizationManager;

public class CustomRoutingDataSource extends AbstractRoutingDataSource {

    @Override
    protected Object determineCurrentLookupKey() {
        return TransactionSynchronizationManager.isCurrentTransactionReadOnly() ? "slave" : "master";
    }
}

File: src/main/java/site/gachontable/infra/database/aurora/CustomRoutingDataSource.kt
package site.gachontable.infra.database.aurora

import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource
import org.springframework.transaction.support.TransactionSynchronizationManager

class CustomRoutingDataSource : AbstractRoutingDataSource() {
    override fun determineCurrentLookupKey(): Any =
        if (TransactionSynchronizationManager.isCurrentTransactionReadOnly())
            SLAVE else MASTER

    companion object {
        private const val MASTER = "master"
        private const val SLAVE = "slave"
    }
}